### ISSUE TITLE
Suggestions for dependencies sometimes ignore rule visibility

### ISSUE THREAD

**Issue by shs96c** [bazelbuild/buildtools#887]
Buildifier tends to suggest a dependency on a target that is later exported by something else and so has limited visibility.

eg. in this example, buildifier should suggest rules depend on `:depend-on-me` and not `:hidden-target`

```py
java_library(
  name = "depend-on-me",
  exports = [":hidden-target"],
  visibility = [ "//visibility:public" ],
)

java_library(
  name = "hidden-target",
  srcs = glob(["*.java"]),
)
```

**Comment by shs96c** [comment]
Obviously, this is a contrived example. More commonly, I use this pattern when migrating maven or gradle projects in a multi-project repo: a top-level target that's public which exports everything within a particular maven project. As the migration to bazel proceeds, I end up with lots of small targets which are aggregated and exported.

**Comment by thomaslangston-mongodb** [comment]
Thinking through how to write this.

target
visibilityOfDep = get visibility of the dep target
if (visibilityOfDep includes target)
  return normal answer

targetsExportingDep = some bazel query to get all targets that export dep
for targetExportingDep in targetsExportingDep
  visibilityOfTargetExportingDep = get visibility of targetExportingDep
  if (visibilityOfTargetExportingDep includes target)
    return targetExportingDep #this only returns the first exporting dep, might be better to return them all

return normal answer with warning that it is not visible
